backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/movement/MoveHUCommand.java

```diff
@@ -0,0 +1,215 @@
+import org.adempiere.exceptions.AdempiereException;
+import org.adempiere.model.PlainContextAware;
+import org.adempiere.warehouse.LocatorId;
+import org.adempiere.warehouse.api.IWarehouseDAO;
+import org.adempiere.warehouse.qrcode.LocatorQRCode;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+public class MoveHUCommand
+{
+	private final ITrxManager trxManager = Services.get(ITrxManager.class);
+	private final IHUMovementBL huMovementBL = Services.get(IHUMovementBL.class);
+	private final IHandlingUnitsBL handlingUnitsBL = Services.get(IHandlingUnitsBL.class);
+	private final IWarehouseDAO warehouseDAO = Services.get(IWarehouseDAO.class);
+	private final HUTransformService huTransformService;
+	private final HUQRCodesService huQRCodesService;
+
+	private final List<HUIdAndQRCode> huIdAndQRCodes;
+	private final GlobalQRCode targetQRCode;
+
+	@Builder
+	private MoveHUCommand(
+			@NonNull final HUQRCodesService huQRCodesService,
+			@NonNull final List<HUIdAndQRCode> husToMove,
+			@NonNull final GlobalQRCode targetQRCode)
+	{
+		this.huQRCodesService = huQRCodesService;
+		this.huTransformService = HUTransformService.builder()
+				.huQRCodesService(huQRCodesService)
+				.build();
+
+		this.huIdAndQRCodes = husToMove;
+		this.targetQRCode = targetQRCode;
+
+	}
+
+	public void execute()
+	{
+		trxManager.runInThreadInheritedTrx(this::executeInTrx);
+	}
+
+	private void executeInTrx()
+	{
+		if (LocatorQRCode.isTypeMatching(targetQRCode))
+		{
+			final LocatorQRCode locatorQRCode = LocatorQRCode.ofGlobalQRCode(targetQRCode);
+
+			moveHUIdsToLocator(extractHUIdsToMove(), locatorQRCode.getLocatorId());
+		}
+		else if (HUQRCode.isTypeMatching(targetQRCode))
+		{
+			getMoveToTargetHUConsumer(HUQRCode.fromGlobalQRCode(targetQRCode))
+					.accept(extractHUIdsToMove());
+		}
+		else
+		{
+			throw new AdempiereException("Move target not handled: " + targetQRCode);
+		}
+	}
+
+	@NonNull
+	private Consumer<List<HuId>> getMoveToTargetHUConsumer(@NonNull final HUQRCode huqrCode)
+	{
+		final I_M_HU targetHU = handlingUnitsBL.getById(huQRCodesService.getHuIdByQRCode(huqrCode));
+		if (handlingUnitsBL.isDestroyed(targetHU))
+		{
+			handlingUnitsBL.setHUStatus(targetHU, PlainContextAware.newWithThreadInheritedTrx(), X_M_HU.HUSTATUS_Active);
+		}
+
+		if (handlingUnitsBL.isLoadingUnit(targetHU))
+		{
+			return getMoveToLUConsumer(targetHU);
+		}
+		else if (handlingUnitsBL.isTransportUnit(targetHU))
+		{
+			return getMoveToTUConsumer(targetHU);
+		}
+		else if (handlingUnitsBL.isVirtual(targetHU))
+		{
+			return getMoveToCUConsumer(targetHU);
+		}
+		else
+		{
+			throw new AdempiereException("Unsupported target!");
+		}
+	}
+
+	@NonNull
+	private Consumer<List<HuId>> getMoveToCUConsumer(final I_M_HU targetHU)
+	{
+		if (!handlingUnitsBL.isVirtual(targetHU))
+		{
+			throw new AdempiereException("Invalid target HU! Expecting CU type, but got=" + handlingUnitsBL.getHU_UnitType(targetHU))
+					.markAsUserValidationError();
```

Comment:
hmm... this is not a user-friendly validation, consider
* having a user friendly translated message if applies,
* or simply don't call `markAsUserValidationError`